### ISSUE TITLE
`Complex` type

### ISSUE THREAD

**Issue by socumbersome** [rafalcieslak/CCoscope#22]
Refers to `typematcher` branch.
Implement built-in `Complex` type, i.e. a thing with double real part and double imaginary part, along with reasonable conversions and operations on it.

**Comment by socumbersome** [comment]
As I started thinking about `Complex` type I realized it poses interesting questions and problems :)
So, currently the code looks like this: cf7c1e9f30bc496bba26513aa308ff26fb4039d4 (I decided to make new `complextype` branch for experiments).

Things more "architecture-related":
- I'm not sure if adding ad-hoc new `ComplexValueAST` is a good idea. I think that when we will have user-defined structs, it will become more reasonable to make `Complex` type reuse `struct` mechanisms (otherwise any "compound" inbuilt type would need yet another AST node -- ugly). But to this moment, let's keep the current approach.
- `print` function seems to be too separated from entities in our compiler (or maybe not `print` function, but the whole `libcco` library?), because I think it would be very convenient to be able to write in our compiler (i.e. in C++) a function that prints objects of type `ExprAST`. It would need to call another print (AKA our current print) to deal with LLVM's entities.
  I think my description doesn't explain anything :D So the example would be:
  In `cco` program we have variable `c` of type `complex` with `Re` part = 4, `Im` part = 2. 
  When we call `print(c)` we expect to see something like `Complex{Re = 4; Im = 2}`.
  Now, how could we achieve it? During codegen of that call expression, we see it's a `print` with argument type `complex`. Then... my (maybe) weird idea is as follows:
1. We codegen `__cco_print_string("Complex{RE =")`
2. We create artificial call `print(Re)` and immediately codegen it (`Re` is a thing of type `Expr`)
3. We codegen `__cco_print_string("; Im = ")`
4. We create artificial call `print(Im)` and immediately codegen it
5. We codegen `__cco_print_string("}")`

I haven't thought it through but I hope it makes some sense.
If I get green light from you then I'll start experimenting with "embedding `libcco` library deeper", so that it knows about `ExprAST`s etc.

---

Now comes implementation-and-bugs-related-story.
When we compile `examples/complex.cco` to LLVM's code we get some weird things:

```
%Complex = type { double, double }

declare void @__cco_print_int(i32)

declare void @__cco_print_double(double)

declare void @__cco_print_bool(i1)

define double @bla(double %Re, double %Im) {
entry:
  %Im_addr3 = alloca double
  %Re_addr2 = alloca double
  %Im_addr = alloca double
  %Re_addr = alloca double
  store double %Re, double* %Re_addr
  store double %Im, double* %Im_addr
  %Re_addr_load = load double, double* %Re_addr
  br i1 true, label %returnBB, label %returnDiscard

returnBB:                                         ; preds = %entry
  ret double %Re_addr_load

returnDiscard:                                    ; preds = %entry
  ret double 0.000000e+00

entry1:                                           ; No predecessors!
  store double %Re, double* %Re_addr2
  store double %Im, double* %Im_addr3
  %Re_addr2_load = load double, double* %Re_addr2
  br i1 true, label %returnBB.4, label %returnDiscard.5

returnBB.4:                                       ; preds = %entry1
  ret double %Re_addr2_load

returnDiscard.5:                                  ; preds = %entry1
  ret double 0.000000e+00
}

define %Complex @newComplex(double %Re, double %Im) {
entry:
  %Im_addr3 = alloca double
  %Re_addr2 = alloca double
  %Im_addr = alloca double
  %Re_addr = alloca double
  store double %Re, double* %Re_addr
  store double %Im, double* %Im_addr
  br i1 true, label %returnBB, label %returnDiscard

returnBB:                                         ; preds = %entry
  ret %Complex { double 1.000000e+00, double 1.000000e+00 }

returnDiscard:                                    ; preds = %entry
  ret %Complex zeroinitializer

entry1:                                           ; No predecessors!
  store double %Re, double* %Re_addr2
  store double %Im, double* %Im_addr3
  br i1 true, label %returnBB.4, label %returnDiscard.5

returnBB.4:                                       ; preds = %entry1
  ret %Complex { double 1.000000e+00, double 1.000000e+00 }

returnDiscard.5:                                  ; preds = %entry1
  ret %Complex zeroinitializer
}

declare %Complex @newComplex.1(double, double)

declare double @bla.2(double, double)

define i32 @main() {
entry:
  %c_addr = alloca %Complex
  %f2_addr = alloca double
  %f1_addr = alloca double
  %i_addr = alloca i32
  store i32 0, i32* %i_addr
  store double 0.000000e+00, double* %f1_addr
  store double 0.000000e+00, double* %f2_addr
  store %Complex zeroinitializer, %Complex* %c_addr
  store double 3.140000e+00, double* %f1_addr
  store double -2.000000e+00, double* %f2_addr
  %f1_addr_load = load double, double* %f1_addr
  %f2_addr_load = load double, double* %f2_addr
  %calltmp = call %Complex @newComplex(double %f1_addr_load, double %f2_addr_load)
  store %Complex %calltmp, %Complex* %c_addr
  %f1_addr_load1 = load double, double* %f1_addr
  call void @__cco_print_double(double %f1_addr_load1)
  %f2_addr_load2 = load double, double* %f2_addr
  call void @__cco_print_double(double %f2_addr_load2)
  ret i32 0
}
```

(`bla` function is declared at `CodegenContext::PrepareStdFunctionPrototypes` just to make sure weird things happen : ))
The way our simple `return` is codegenned is somewhat crazy, having duplicated half of the body. And also, implementation of `newComplex` is buggy as it is insensitive to its arguments whatsoever.

**Comment by rafalcieslak** [comment]
I'm not sure I understand what the problem with `print`ing complex types is. If you wish, you can totally link the compiler with the runtime library (when building compiler executable), it does make sense. But the other way round (linking the runtime library with the compiler when building the runtime library) is dangerous - the runtime library is statically linked when generating binaries (from a `cco` file), so we might end up generating binaries that contain parts of the compiler...

Also, since we consider `Complex` a built-in type, why won't we just create a `__cco_print_complex(void*)`? The implementation of `__cco_print_complex` would print the brackets and extract the real and imaginary value from the pointer it received. When generating code for `print`ing a `Complex` we would just emit a call to `__cco_print_complex` with a pointer to the Complex value we wish to print.

What am I missing here?

**Comment by socumbersome** [comment]
Yes, I also feared the possibility of generating binaries with compiler parts, but after some time I kind of got used to that idea and... now I don't really see what would be inappropriate in this solution? It's supposed to be runtime library after all.

I see two problems with `void*` solution:
- In order to be able to extract `Re` and `Im`, `__cco_print_complex` would need to know the `complex` type, so doesn't that mean more or less the same as linking runtime library when building it (exactly what you marked as dangerous)?
- Things stored in `complex` can be arbitrary expressions that evaluate to `double`, so I can't imagine how `__cco_print_complex` would be able to print things of arbitrary type from our `AST` hierarchy (or `llvm::Value*` things if we for some reason decide that printing should happen after codegen).
- I don't like `void*` ; )

But I have some vague idea as to how I would see that -- we probably need some `dump` methods for each AST that would act almost as `print` ("strongly-embedded" runtime library I'd say), but which use `print` from `libcco` when needed. Well, I'll try to conduct some experiment.

**Comment by rafalcieslak** [comment]
What I dislike about embedding compiler sources within the runtime library is that, well in most cases it is fine for the library to understand the concepts used by the compiler, but sometimes it would be very awkward. For example, I don't suppose the generated binary should contain any information about the nature of AST, nor any functions that perform operations on it. Similarly, it should not contain any code that uses llvm to generate some instructions, that's just wicked.

Furthermore, I worry that if we link the generated binary with a runtime library that was created from both compiler and library sources, then such generated binary would need to be dynamically linked with llvm libraries, and thus all binaries our compiler creates would have llvm as dependency. This is naturally bad.

I am not sure the problems you listed are valid:
- First, in order to understand `complex` type, the runtime library requires just the header files from compiler source. That is totally fine, as it is the same as exposing the API. We might even want to create a _third_ code module, a library of common operations that are useful to both compiler and the standard library (such library would keep external dependencies to minimum and should in no way know about llvm).
- Why would we ever want to call `print_complex` on anything that is not an actual value of a complex number? If we did, then we actually require two different printers - one for printing a value, and another for printing an expression tree (the first goes into the runtime library, the other into the compiler, this is not code duplication because they do significantly different stuff, it just happens to produce a similar output, sometimes).

**Comment by socumbersome** [comment]
Ok, at https://github.com/rafalcieslak/CCoscope/commit/77e39e06ccee94021d34aaf434f1806ebfdc5390 I have a ready implementation (at least for printing -- haven't tested arithmetic yet). I've tried to make it so that `libcco` knows nothing about our compiler's types and I think it is not even as awkward as I had thought it would be. The compiler just extracts two doubles from our `complex` type and then let's `libcco` do the rest of the job.
But even though it works, I'm not entirely satisfied with this solution, because when we have real `structs` in our language, it will become more difficult to print things inside them (imagine struct with a field that is a struct etc.). We certainly cannot afford creating millions of `__cco_print_xxx` functions so I consider this a temporary solution (or, even if we stay with it, we'll have to think of something better when `structs` come into play).

**Comment by rafalcieslak** [comment]
I am unable to test these changes at 8b8ed92e462d4a5ed9bdaf7ec6dcb7fa85136d25 as compilation fails for me with [these errors](http://pastebin.com/raw/K38MF6X8) (basically, `llvm::StructType*` fails to convert to `llvm::Value*` - according to the [documentation](http://llvm.org/docs/doxygen/html/classllvm_1_1StructType.html) it does make sense).

It appears that `CreateStructGEP` prototype has changed between llvm 3.5 and 3.8. Is this something we can fix, or shall I - finally - switch to 3.8?

**Comment by socumbersome** [comment]
Well... I guess there may be many more differences between 3.5 and 3.8 and I think it will be better to adjust everything "upward". So if this is not very inconvenient for you, I'd suggest moving to version 3.8.
